File: app/src/main/java/com/yibao/music/fragment/AlbumFragment.java
package com.yibao.music.fragment;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.viewpager.widget.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.yibao.music.R;
import com.yibao.music.adapter.AlbumCategoryPagerAdapter;
import com.yibao.music.adapter.DetailsViewAdapter;
import com.yibao.music.base.BaseMusicFragment;
import com.yibao.music.base.listener.MusicPagerListener;
import com.yibao.music.fragment.dialogfrag.MoreMenuBottomDialog;
import com.yibao.music.model.AlbumInfo;
import com.yibao.music.aidl.MusicBean;
import com.yibao.music.model.greendao.MusicBeanDao;
import com.yibao.music.util.ColorUtil;
import com.yibao.music.util.Constants;
import com.yibao.music.view.music.DetailsView;
import com.yibao.music.view.music.MusicToolBar;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;


/**
 * @项目名： ArtisanMusic
 * @包名： com.yibao.music.album
 * @文件名: AlbumFragment
 * @author: Stran
 * @Email: www.strangermy@outlook.com / www.stranger98@gmail.com
 * @创建时间: 2018/2/8 20:01
 * @描述： {TODO}
 */

public class AlbumFragment extends BaseMusicFragment {
    @BindView(R.id.music_toolbar_list)
    MusicToolBar mMusicToolBar;
    @BindView(R.id.iv_album_category_random_paly)
    ImageView mIvAlbumCategoryRandomPaly;
    @BindView(R.id.iv_album_category_list)
    ImageView mIvAlbumCategoryList;
    @BindView(R.id.tv_album_category_list)
    TextView mTvAlbumCategoryList;
    @BindView(R.id.album_category_list_ll)
    LinearLayout mAlbumCategoryListLl;
    @BindView(R.id.iv_album_category_tile)
    ImageView mIvAlbumCategoryTile;
    @BindView(R.id.tv_album_category_tile)
    TextView mTvAlbumCategoryTile;
    @BindView(R.id.album_category_tile_ll)
    LinearLayout mAlbumCategoryTileLl;
    @BindView(R.id.iv_album_category_paly)
    ImageView mIvAlbumCategoryPaly;

    @BindView(R.id.view_pager_album)
    ViewPager mViewPager;
    @BindView(R.id.details_view)
    DetailsView mDetailsView;
    @BindView(R.id.album_content_view)
    LinearLayout mAlbumContentView;

    private DetailsViewAdapter mDetailsAdapter;
    private boolean isShowDetailsView = false;
    private boolean mDetailViewFlag = true;
    private List<MusicBean> mDetailList;
    private String detailsViewTitle;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.album_fragment, container, false);
        unbinder = ButterKnife.bind(this, view);
        initData();
        return view;
    }


    private void initData() {
        AlbumCategoryPagerAdapter pagerAdapter = new AlbumCategoryPagerAdapter(getChildFragmentManager());
        mViewPager.setAdapter(pagerAdapter);
        mViewPager.addOnPageChangeListener(new MusicPagerListener() {
            @Override
            public void onPageSelected(int position) {
                switchCategory(position);
            }
        });
        initRxbusData();
        initListener();
    }


    @Override
    public void onResume() {
        super.onResume();
        mMusicToolBar.setToolbarTitle(isShowDetailsView ? detailsViewTitle : getString(R.string.music_album));
        initRxBusData();
    }

    @Override
    protected boolean getIsOpenDetail() {
        return isShowDetailsView;
    }

    private void initRxBusData() {
        disposeToolbar();
        if (mEditDisposable == null) {
            mEditDisposable = mBus.toObservableType(Constants.FRAGMENT_ALBUM, Object.class).subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread()).subscribe(editBean -> {
                        mMusicToolBar.setTvEditText(R.string.tv_edit);
                        mMusicToolBar.setTvDeleteVisibility(View.GONE);
                        isShowDetailsView = false;
                    });
        }
    }

    private void initListener() {
        mMusicToolBar.setClickListener(new MusicToolBar.OnToolbarClickListener() {
            @Override
            public void clickEdit() {
                if (mDetailViewFlag) {
                    mBus.post(Constants.ALBUM_FAG_EDIT, Constants.NUMBER_THRRE);
                }
                mMusicToolBar.setTvDeleteVisibility(isShowDetailsView ? View.GONE : View.VISIBLE);
                mMusicToolBar.setTvEditText(!isShowDetailsView ? R.string.tv_edit : R.string.complete);
                showDetailsView(null);
            }

            @Override
            public void switchMusicControlBar() {
                switchControlBar();
            }

            @Override
            public void clickDelete() {
                mBus.post(Constants.ALBUM_FAG_EDIT, Constants.NUMBER_FOUR);
            }
        });
    }

    private void initRxbusData() {
        mCompositeDisposable.add(mBus.toObserverable(AlbumInfo.class)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::showDetailsView));
    }

    @OnClick({R.id.iv_album_category_random_paly,
            R.id.album_category_list_ll, R.id.album_category_tile_ll, R.id.iv_album_category_paly})
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.iv_album_category_random_paly:
            case R.id.iv_album_category_paly:
                randomPlayMusic();
                break;
            case R.id.album_category_list_ll:
                switchCategory(Constants.NUMBER_ZERO);
                break;
            case R.id.album_category_tile_ll:
                switchCategory(Constants.NUMBER_ONE);
                break;
            default:
                break;
        }
    }

    private void showDetailsView(AlbumInfo albumInfo) {
        if (isShowDetailsView) {
            mDetailsView.setVisibility(View.GONE);
            mMusicToolBar.setToolbarTitle(getString(R.string.music_album));
            mDetailViewFlag = true;
        } else {
            if (albumInfo != null) {
                mDetailViewFlag = false;
                mDetailsView.setVisibility(View.VISIBLE);
                mDetailList = mMusicBeanDao.queryBuilder().where(MusicBeanDao.Properties.Album.eq(albumInfo.getAlbumName())).build().list();
                // DetailsView播放音乐需要的参数
                mDetailsView.setDataFlag(mFragmentManager, mDetailList.size(), albumInfo.getAlbumName(), Constants.NUMBER_TWO);
                mDetailsAdapter = new DetailsViewAdapter(mActivity, mDetailList, Constants.NUMBER_TWO);
                mDetailsView.setAdapter(Constants.NUMBER_TWO, albumInfo, mDetailsAdapter);
                mDetailsAdapter.setOnItemMenuListener((int position, MusicBean musicBean) ->
                        MoreMenuBottomDialog.newInstance(musicBean, position, false, false).getBottomDialog(mActivity));
                interceptBackEvent(Constants.NUMBER_TWELVE);
                detailsViewTitle = albumInfo.getAlbumName();
                mMusicToolBar.setToolbarTitle(detailsViewTitle);
            }
        }
        mMusicToolBar.setTvEditText(isShowDetailsView ? R.string.tv_edit : R.string.back);
        isShowDetailsView = !isShowDetailsView;
    }

    @Override
    protected void deleteItem(int musicPosition) {
        super.deleteItem(musicPosition);
        if (mDetailList != null && mDetailsAdapter != null) {
            mDetailList.remove(musicPosition);
            mDetailsAdapter.setData(mDetailList);
        }
    }

    @Override
    protected void handleDetailsBack(int detailFlag) {
        if (detailFlag == Constants.NUMBER_TWELVE) {
            showDetailsView(null);
        }
    }


    private void switchCategory(int showType) {
        mViewPager.setCurrentItem(showType, false);
        if (showType == Constants.NUMBER_ZERO) {

            mAlbumCategoryListLl.setBackgroundResource(R.drawable.btn_category_songname_down_selector);
            mIvAlbumCategoryList.setImageResource(R.drawable.album_category_list_down_selector);
            mTvAlbumCategoryList.setTextColor(ColorUtil.wihtle);


            mAlbumCategoryTileLl.setBackgroundResource(R.drawable.btn_category_views_selector);

            mIvAlbumCategoryTile.setImageResource(R.drawable.album_category_tile_selector);

            mTvAlbumCategoryTile.setTextColor(ColorUtil.textName);

        } else if (showType == Constants.NUMBER_ONE) {

            mAlbumCategoryTileLl.setBackgroundResource(R.drawable.btn_category_views_down_selector);
            mIvAlbumCategoryTile.setImageResource(R.drawable.album_category_tile_down_selector);
            mTvAlbumCategoryTile.setTextColor(ColorUtil.wihtle);

            mAlbumCategoryListLl.setBackgroundResource(R.drawable.btn_category_songname_selector);
            mIvAlbumCategoryList.setImageResource(R.drawable.album_category_list_selector);
            mTvAlbumCategoryList.setTextColor(ColorUtil.textName);


        }


    }


    public static AlbumFragment newInstance() {
        return new AlbumFragment();
    }


}

File: app/src/main/java/com/yibao/music/fragment/AlbumFragment.kt
package com.yibao.music.fragment

import android.view.View
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
import com.yibao.music.R
import com.yibao.music.adapter.AlbumViewPagerAdapter
import com.yibao.music.adapter.DetailsViewAdapter
import com.yibao.music.base.bindings.BaseBindingAdapter
import com.yibao.music.base.bindings.BaseMusicFragmentDev
import com.yibao.music.databinding.AlbumFragmentBinding
import com.yibao.music.fragment.dialogfrag.MoreMenuBottomDialog
import com.yibao.music.model.AlbumInfo
import com.yibao.music.model.MusicBean
import com.yibao.music.model.greendao.MusicBeanDao
import com.yibao.music.util.ColorUtil
import com.yibao.music.util.Constant
import com.yibao.music.util.MusicListUtil
import com.yibao.music.view.music.MusicToolBar.OnToolbarClickListener
import com.yibao.music.viewmodel.AlbumViewModel

/**
 * @项目名： ArtisanMusic
 * @包名： com.yibao.music.album
 * @文件名: AlbumFragment
 * @author: Stran
 * @Email: www.strangermy@outlook.com / www.stranger98@gmail.com
 * @创建时间: 2018/2/8 20:01
 * @描述： {TODO}
 */
class AlbumFragment : BaseMusicFragmentDev<AlbumFragmentBinding>(), View.OnClickListener,
    SwipeRefreshLayout.OnRefreshListener {
    private val mViewModel: AlbumViewModel by lazy { gets(AlbumViewModel::class.java) }
    private var mDetailsAdapter: DetailsViewAdapter? = null
    private var isShowDetailsView = false
    private var mDetailList = ArrayList<MusicBean>()
    override fun initView() {
        mBinding.musicBar.setToolbarTitle(getString(R.string.music_album))
        mBinding.musicBar.isShowAlbumWall(visibility = true)
    }


    override fun initData() {
        val pagerAdapter = AlbumViewPagerAdapter(this, mViewModel)
        mBinding.viewPager2Album.adapter = pagerAdapter
        mBinding.viewPager2Album.isUserInputEnabled = false
        mBinding.viewPager2Album.registerOnPageChangeCallback(object : OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                switchCategory(position)
            }
        })

        initListener()
    }

    override fun onResume() {
        super.onResume()
        //    接收AlbumAdapter发过来的当前点击Item的Position
        mViewModel.albumViewModel.observe(this) { bean ->
            showDetailsView(bean)
        }

    }


    private fun initListener() {

        mBinding.albumCategory.ivAlbumCategoryRandomPlay.setOnClickListener(this)
        mBinding.albumCategory.albumCategoryTileLl.setOnClickListener(this)
        mBinding.albumCategory.albumCategoryListLl.setOnClickListener(this)
        mBinding.albumCategory.ivAlbumCategoryPlay.setOnClickListener(this)

        mBinding.musicBar.setClickListener(object : OnToolbarClickListener {
            override fun clickEdit() {
                if (isShowDetailsView) {
                    showDetailsView(null)
                }
            }

            override fun switchMusicControlBar() {
                switchControlBar()
            }

            override fun clickDelete() {}
        })
    }


    override fun onClick(v: View) {
        when (v.id) {
            R.id.iv_album_category_random_play, R.id.iv_album_category_play -> randomPlayMusic(6)
            R.id.album_category_list_ll -> switchCategory(Constant.NUMBER_ZERO)
            R.id.album_category_tile_ll -> switchCategory(Constant.NUMBER_ONE)
            else -> {}
        }
    }

    private fun showDetailsView(albumInfo: AlbumInfo?) {
        if (isShowDetailsView) {
            mBinding.detailsView.visibility = View.GONE
            mBinding.musicBar.setToolbarTitle(getString(R.string.music_album))
            mBinding.musicBar.setTvEditVisibility(false)
        } else {
            if (albumInfo != null) {
                mBinding.detailsView.visibility = View.VISIBLE
                mBinding.musicBar.setTvEditVisibility(true)
                mBinding.musicBar.setToolbarTitle(albumInfo.albumName)
                mBinding.musicBar.setTvEditText(R.string.music_album)

                mDetailList = MusicListUtil.sortByAbc(
                    mMusicBeanDao.queryBuilder()
                        .where(MusicBeanDao.Properties.Album.eq(albumInfo.albumName)).build()
                        .list()
                ) as ArrayList<MusicBean>
                // DetailsView播放音乐需要的参数
                mBinding.detailsView.setDataFlag(
                    childFragmentManager,
                    mDetailList.size,
                    albumInfo.albumName,
                    Constant.NUMBER_SEVEN
                )
                mDetailsAdapter = DetailsViewAdapter(
                    mContext,
                    mDetailList,
                    Constant.NUMBER_SEVEN,
                    albumInfo.albumName
                )
                mBinding.detailsView.setAdapter(Constant.NUMBER_TWO, albumInfo, mDetailsAdapter)
                mDetailsAdapter!!.setOnItemMenuListener(object :
                    BaseBindingAdapter.OnOpenItemMoreMenuListener {
                    override fun openClickMoreMenu(position: Int, musicBean: MusicBean) {
                        MoreMenuBottomDialog.newInstance(
                            musicBean,
                            position,
                            false,
                            false
                        ).getBottomDialog(mActivity)
                    }
                })
            }
        }
        isShowDetailsView = !isShowDetailsView

    }

    override fun deleteItem(musicPosition: Int) {
        super.deleteItem(musicPosition)
        if (mDetailsAdapter != null) {
            mDetailList.removeAt(musicPosition)
            mDetailsAdapter!!.setData(mDetailList)
        }
    }


    private fun switchCategory(showType: Int) {
        mBinding.viewPager2Album.setCurrentItem(showType, false)
        if (showType == Constant.NUMBER_ZERO) {
            mBinding.albumCategory.albumCategoryListLl.setBackgroundResource(R.drawable.btn_category_start_down_selector)
            mBinding.albumCategory.ivAlbumCategoryList.setImageResource(R.drawable.album_category_list_down_selector)
            mBinding.albumCategory.tvAlbumCategoryList.setTextColor(ColorUtil.wihtle)
            mBinding.albumCategory.albumCategoryTileLl.setBackgroundResource(R.drawable.btn_category_end_selector)
            mBinding.albumCategory.ivAlbumCategoryTile.setImageResource(R.drawable.album_category_tile_selector)
            mBinding.albumCategory.tvAlbumCategoryTile.setTextColor(ColorUtil.textName)
        } else if (showType == Constant.NUMBER_ONE) {
            mBinding.albumCategory.albumCategoryTileLl.setBackgroundResource(R.drawable.btn_category_end_down_selector)
            mBinding.albumCategory.ivAlbumCategoryTile.setImageResource(R.drawable.album_category_tile_down_selector)
            mBinding.albumCategory.tvAlbumCategoryTile.setTextColor(ColorUtil.wihtle)
            mBinding.albumCategory.albumCategoryListLl.setBackgroundResource(R.drawable.btn_category_start_selector)
            mBinding.albumCategory.ivAlbumCategoryList.setImageResource(R.drawable.album_category_list_selector)
            mBinding.albumCategory.tvAlbumCategoryList.setTextColor(ColorUtil.textName)
        }
    }

    companion object {
        fun newInstance(): AlbumFragment {
            return AlbumFragment()
        }
    }


    override fun onBackPressed(): Boolean {
        return if (isShowDetailsView && isVisible && isResumed) {
            showDetailsView(null)
            true
        } else {
            false
        }
    }

    override fun onRefresh() {
        showDetailsView(null)

    }
}